### PR TITLE
CSR condition status, lastTransitionTime, versioned validation

### LINKED ISSUE #90510: csr approved than denied ，can not be approved again
**What happened**:
```
root@master01 12certificate]# kubectl certificate approve  my-svc.my-namespace --force
certificatesigningrequest.certificates.k8s.io/my-svc.my-namespace approved
[root@master01 12certificate]# kubectl get csr
NAME                  AGE   REQUESTOR   CONDITION
my-svc.my-namespace   19m   admin       Denied,Issued
[root@master01 12certificate]# ^C
[root@master01 12certificate]# ^C
[root@master01 12certificate]# kubectl get csr
NAME                  AGE   REQUESTOR   CONDITION
my-svc.my-namespace   27m   admin       Denied,Issued
```
**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
```
[root@master01 12certificate]# kubectl version
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:18:23Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T19:09:08Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by liggitt**
**What type of PR is this?**
/kind api-change
/kind feature

**What this PR does / why we need it**:

Implements v1beta1 portions of the [CSR KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190607-certificates-api.md) (https://github.com/kubernetes/enhancements/issues/1513)

Best reviewed commit-by-commit

* Adds a well-known condition type of `Failed` for use by signers that want to indicate permanent failure
  * Updates the signer controller to record a Failed condition when a request does not pass signerName validation
  * Updates printers/describers to surface failed status
  * Updates cert manager to stop waiting for failed CSRs
  * Updates cert cleaner to clean up stale failed CSRs
  * Allows conditions other than `Approved` and `Denied` to be added via `/status`
* Brings CSR conditions in line with recommended condition schema:
  * Adds a status field (for compatibility, defaults to `True` if unspecified, may only be `True` for `Approved`, `Denied`, and `Failed` conditions)
  * Adds a lastTransitionTime field
* Plumbs API version information to CSR validation
* Partitions validation functions for the main resource, status, and approval subresources
* In all API versions:
  * `Approved` and `Denied` conditions are mutually exclusive (fixes https://github.com/kubernetes/kubernetes/issues/90510)
  * `Approved`, `Denied`, and `Failed` conditions cannot be removed once set
* In versions other than v1beta1:
  * `status.certificate` is required to be well-formed
  * `status.certificate` is immutable once set 
  * duplicate condition types are not allowed
  * empty condition type is not allowed
  * legacy-unknown signer is not allowed

**Does this PR introduce a user-facing change?**:
```release-note
CertificateSigningRequest API conditions were updated:
* a `status` field was added; this field defaults to `True`, and may only be set to `True` for `Approved`, `Denied`, and `Failed` conditions
* a `lastTransitionTime` field was added
* a `Failed` condition type was added to allow signers to indicate permanent failure; this condition can be added via the `certificatesigningrequests/status` subresource.
* `Approved` and `Denied` conditions are mutually exclusive
* `Approved`, `Denied`, and `Failed` conditions can no longer be removed from a CSR
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:
```docs
https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/20190607-certificates-api.md
```

/sig auth
/milestone v1.19
/priority important-soon
/cc @deads2k @munnerz

### CODE REVIEW

File: pkg/apis/certificates/types.go [hunk 1]
```diff
@@ -72,6 +75,28 @@ type CertificateSigningRequestSpec struct {
...
+
+	// Signs serving certificates that are honored as a valid kubelet serving
+	// certificate by the kube-apiserver, but has no other guarantees.
+	KubeletServingSignerName = "kubernetes.io/kubelet-serving"
+
+	// Has no guarantees for trust at all. Some distributions may honor these
+	// as client certs, but that behavior is not standard kubernetes behavior.
+	LegacyUnknownSignerName = "kubernetes.io/legacy-unknown"
+)
+
```
[the remark this answers is not in the document]
Comment (munnerz) [reply pkg/apis/certificates/types.go:99]: I'd have thought they only need defining here if they are actually used as part of conversion? But you know better than I do with this stuff 😄

File: pkg/apis/certificates/v1beta1/defaults.go [hunk 2]
```diff
@@ -18,10 +18,12 @@ package v1beta1
 
 import (
 	"crypto/x509"
+	"fmt"
 	"reflect"
 	"strings"
 
 	certificatesv1beta1 "k8s.io/api/certificates/v1beta1"
+	v1 "k8s.io/api/core/v1"
```
Comment (munnerz) [inline pkg/apis/certificates/v1beta1/defaults.go:26]: nit: don't need `v1` alias

File: pkg/apis/certificates/validation/validation.go [hunk 3]
```diff
@@ -17,16 +17,44 @@ limitations under the License.
...
+type certificateValidationOptions struct {
+	// the API version used to make the request being validated
+	version schema.GroupVersion
+
+	// The following allow modifications only permitted via certain update paths
+
+	// allow adding Approved/Denied conditions
+	allowSettingApproval bool
+	// allow modifying status.certificate
+	allowSettingCertificate bool
```
[the remark this answers is not in the document]
Comment (munnerz) [reply pkg/apis/certificates/validation/validation.go:41]: Ah, I guess this is so one validator structure can be used for all of 'status', 'approval' and the root resource?

File: pkg/apis/certificates/validation/validation.go [hunk 4]
```diff
@@ -140,8 +196,128 @@ func ValidateCertificateSigningRequestSignerName(fldPath *field.Path, signerName
...
+		return true // compatibility with existing data
+	default:
+		return false
+	}
+}
+
+func allowDuplicateConditionTypes(version schema.GroupVersion, oldCSR *certificates.CertificateSigningRequest) bool {
+	switch {
+	case version == certificatesv1beta1.SchemeGroupVersion:
+		return true // compatibility with v1beta1
```
Comment (munnerz) [inline pkg/apis/certificates/validation/validation.go:288]: github doesn't have a wide enough pallet of emojis for me to properly express my reaction here

File: pkg/apis/certificates/validation/validation.go [hunk 5]
```diff
@@ -140,8 +196,128 @@ func ValidateCertificateSigningRequestSignerName(fldPath *field.Path, signerName
...
+	case oldCSR != nil && hasDuplicateConditionTypes(oldCSR):
+		return true // compatibility with existing data
+	default:
+		return false
+	}
+}
+func hasDuplicateConditionTypes(csr *certificates.CertificateSigningRequest) bool {
+	seen := map[string]bool{}
```
Comment (munnerz) [inline pkg/apis/certificates/validation/validation.go:296]: nit: we could just use a `ConditionType` here and avoid the casting?

File: pkg/apis/certificates/validation/validation.go [hunk 6]
```diff
@@ -140,8 +196,128 @@ func ValidateCertificateSigningRequestSignerName(fldPath *field.Path, signerName
...
+		}
+		seen[string(c.Type)] = true
+	}
+	return false
+}
+
+func allowEmptyConditionType(version schema.GroupVersion, oldCSR *certificates.CertificateSigningRequest) bool {
+	switch {
+	case version == certificatesv1beta1.SchemeGroupVersion:
+		return true // compatibility with v1beta1
```
Comment (munnerz) [inline pkg/apis/certificates/validation/validation.go:309]: 🤦

File: pkg/apis/certificates/validation/validation_test.go [hunk 7]
```diff
@@ -261,7 +263,7 @@ func TestValidateCertificateSigningRequest(t *testing.T) {
 	}
 	for name, test := range tests {
 		t.Run(name, func(t *testing.T) {
-			el := ValidateCertificateSigningRequest(&test.csr)
+			el := ValidateCertificateSigningRequest(&test.csr, test.gv)
```
Comment (munnerz) [inline pkg/apis/certificates/validation/validation_test.go:266]: (as a marker/reminder for me for later): needs more tests 😅

File: staging/src/k8s.io/api/certificates/v1beta1/types.go [hunk 8]
```diff
@@ -132,11 +133,16 @@ type RequestConditionType string
...
 )
 
 type CertificateSigningRequestCondition struct {
-	// request approval state, currently Approved or Denied.
+	// type of the condition. Known conditions include "Approved", "Denied", and "Failed".
 	Type RequestConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=RequestConditionType"`
+	// Status of the condition, one of True, False, Unknown.
+	// Approved, Denied, and Failed conditions may not be "False" or "Unknown".
+	// If unset, should be treated as True.
+	Status v1.ConditionStatus `json:"status" protobuf:"bytes,5,opt,name=status,casttype=k8s.io/api/core/v1.ConditionStatus"`
```
Comment (munnerz) [inline staging/src/k8s.io/api/certificates/v1beta1/types.go:145]: This should have `// +optional`

File: staging/src/k8s.io/api/certificates/v1beta1/types.go [hunk 9]
```diff
@@ -132,11 +133,16 @@ type RequestConditionType string
 const (
 	CertificateApproved RequestConditionType = "Approved"
 	CertificateDenied   RequestConditionType = "Denied"
+	CertificateFailed   RequestConditionType = "Failed"
 )
 
 type CertificateSigningRequestCondition struct {
-	// request approval state, currently Approved or Denied.
+	// type of the condition. Known conditions include "Approved", "Denied", and "Failed".
 	Type RequestConditionType `json:"type" protobuf:"bytes,1,opt,name=type,casttype=RequestConditionType"`
```
Comment (munnerz) [inline staging/src/k8s.io/api/certificates/v1beta1/types.go:141]: :eyes: we don't have `// +optional` here - doesn't this mean this field is already required with newer kubectl versions? (but I guess not with API clients)

File: pkg/apis/certificates/types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/apis/certificates/types.go:99]: yes, validation for v1 prevents creation of CSRs with legacy unknown signer and uses the internal API

File: pkg/apis/certificates/validation/validation.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/apis/certificates/validation/validation.go:41]: right... when we call from ValidateStatusUpdate, we allow setting the certificate. When we call from ValidateApprovalUpdate, we don't.

File: staging/src/k8s.io/api/certificates/v1beta1/types.go (hunk 9 above)
Comment (liggitt) [reply staging/src/k8s.io/api/certificates/v1beta1/types.go:141]: this hasn't changed, so... yes... anything submitting a condition using openapi validation would need to include the field, but they could set it's value to an empty string

File: pkg/apis/certificates/v1beta1/defaults.go (hunk 2 above)
Comment (liggitt) [reply pkg/apis/certificates/v1beta1/defaults.go:26]: goimports adds it automatically

File: pkg/apis/certificates/v1beta1/defaults.go [hunk 10]
```diff
@@ -59,18 +67,34 @@ func DefaultSignerNameFromSpec(obj *certificatesv1beta1.CertificateSigningReques
...
+	uriSANNotAllowedErr           = fmt.Errorf("URI subjectAltNames are not allowed")
+)
+
 func IsKubeletServingCSR(req *x509.CertificateRequest, usages []certificatesv1beta1.KeyUsage) bool {
+	return ValidateKubeletServingCSR(req, usages) == nil
+}
+func ValidateKubeletServingCSR(req *x509.CertificateRequest, usages []certificatesv1beta1.KeyUsage) error {
 	if !reflect.DeepEqual([]string{"system:nodes"}, req.Subject.Organization) {
-		return false
+		return organizationNotSystemNodesErr
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/apis/certificates/v1beta1/defaults.go:85]: I think I'll just stick with the first one we hit for now

File: pkg/apis/certificates/validation/validation_test.go (hunk 7 above)
Comment (liggitt) [reply pkg/apis/certificates/validation/validation_test.go]: so many tests added

**Comment by liggitt** [comment]
this is ready for another pass

File: pkg/controller/certificates/cleaner/cleaner.go [hunk 11]
```diff
@@ -158,6 +159,19 @@ func isDeniedPastDeadline(csr *capi.CertificateSigningRequest) bool {
 	return false
 }
 
+// isFailedPastDeadline checks if the certificate has a Failed status and the
+// creation time of the CSR is passed the deadline that pending requests are
+// maintained for.
+func isFailedPastDeadline(csr *capi.CertificateSigningRequest) bool {
```
Comment (munnerz) [inline pkg/controller/certificates/cleaner/cleaner.go:165]: Despite this being quite simple, might be worth including a unit test here given there's already plumbing for it? https://github.com/kubernetes/kubernetes/blob/dcabb7e6a81e7e60a43a82a380bf01fd0604b8f7/pkg/controller/certificates/cleaner/cleaner_test.go

Unless there's an integration test I've missed 😄

File: staging/src/k8s.io/client-go/util/certificate/csr/csr.go [hunk 12]
```diff
@@ -112,17 +112,24 @@ func WaitForCertificate(ctx context.Context, client certificatesclient.Certifica
...
-					klog.V(2).Infof("certificate signing request %s is approved, waiting to be issued", csr.Name)
+					approved = true
+				}
+			}
+			if approved {
+				if csr.Status.Certificate != nil {
+					klog.V(2).Infof("certificate signing request %s is issued", csr.Name)
+					return true, nil
 				}
+				klog.V(2).Infof("certificate signing request %s is approved, waiting to be issued", csr.Name)
```
Comment (munnerz) [inline staging/src/k8s.io/client-go/util/certificate/csr/csr.go:132]: 😢 there are no tests for this function yet.. https://github.com/kubernetes/kubernetes/blob/a333e591b8e1bca4692eca136389de621979c5ed/staging/src/k8s.io/client-go/util/certificate/csr/csr_test.go

Probably a good thing that can be done in a follow up

File: pkg/apis/certificates/types.go [hunk 13]
```diff
@@ -91,11 +116,17 @@ type RequestConditionType string
...
+	CertificateFailed   RequestConditionType = "Failed"
 )
 
 type CertificateSigningRequestCondition struct {
-	// request approval state, currently Approved or Denied.
+	// type of the condition. Known conditions include "Approved", "Denied", and "Failed".
 	Type RequestConditionType
+	// Status of the condition, one of True, False, Unknown.
+	// Approved, Denied, and Failed conditions may not be "False" or "Unknown".
+	// If unset, should be treated as True.
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/apis/certificates/types.go:127]: The docs are also for clients, which can easily read the docs from the latest version and run against an N-1 server.

File: staging/src/k8s.io/api/certificates/v1beta1/types.go [hunk 14]
```diff
@@ -134,11 +135,17 @@ type RequestConditionType string
...
+	// Status of the condition, one of True, False, Unknown.
+	// Approved, Denied, and Failed conditions may not be "False" or "Unknown".
+	// If unset, should be treated as True.
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/api/certificates/v1beta1/types.go:146]: the new client / old server was what I had in mind. I'll also include the default.

File: staging/src/k8s.io/client-go/util/certificate/csr/csr.go (hunk 12 above)
Comment (liggitt) [reply staging/src/k8s.io/client-go/util/certificate/csr/csr.go:132]: the kubelet cert bootstrap unit tests exercise it

File: pkg/controller/certificates/cleaner/cleaner.go (hunk 11 above)
Comment (liggitt) [reply pkg/controller/certificates/cleaner/cleaner.go:165]: added a unit test

File: pkg/apis/certificates/validation/validation.go [hunk 15]
```diff
@@ -60,7 +152,71 @@ func ValidateCertificateSigningRequest(csr *certificates.CertificateSigningReque
...
+	if !opts.allowLegacySignerName && csr.Spec.SignerName == certificates.LegacyUnknownSignerName {
+		allErrs = append(allErrs, field.Invalid(specPath.Child("signerName"), csr.Spec.SignerName, "the legacy signerName is not allowed via this API version"))
+	} else {
+		allErrs = append(allErrs, ValidateCertificateSigningRequestSignerName(specPath.Child("signerName"), csr.Spec.SignerName)...)
+	}
+	allErrs = append(allErrs, validateConditions(field.NewPath("status", "conditions"), csr, opts)...)
+
+	if !opts.allowArbitraryCertificate {
+		if err := validateCertificate(csr.Status.Certificate); err != nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("status", "certificate"), "", err.Error()))
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/apis/certificates/validation/validation.go:164]: not sure.... when we redact, we've done the following:

```
allErrs = append(allErrs, field.Invalid(dataPath.Key(core.DockerConfigKey), "<secret contents redacted>", err.Error()))
allErrs = append(allErrs, field.Invalid(dataPath.Key(core.DockerConfigJSONKey), "<secret contents redacted>", err.Error()))
```

for things that are impractical to include the entire serialized struct, we've typically just done "":
```
return append(allErrs, field.Invalid(fldPath, "", "cannot set both resource and service backends"))
allErrs = append(allErrs, field.Invalid(fldPath.Child("openAPIV3Schema"), "", err.Error()))
allErrs = append(allErrs, field.Invalid(fldPath, "", fmt.Sprintf("error building validator: %v", err)))
```

File: staging/src/k8s.io/api/certificates/v1beta1/types.go [hunk 16]
```diff
@@ -148,6 +155,11 @@ type CertificateSigningRequestCondition struct {
 	// timestamp for the last update to this condition
 	// +optional
 	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty" protobuf:"bytes,4,opt,name=lastUpdateTime"`
+	// lastTransitionTime is the time the condition last transitioned from one status to another.
+	// If unset, when a new condition type is added or an existing condition's status is changed,
+	// the server defaults this to the current time.
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/api/certificates/v1beta1/types.go:161]: probably not? we don't do it for other conditions... the certificates API was somewhat unique in its use of conditions, and this is matching the existing timestamp defaulting for lastUpdateTime

File: pkg/apis/certificates/validation/validation.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/apis/certificates/validation/validation.go]: set to `"<certificate data>"` to make errors more legible

**Comment by liggitt** [comment]
Comments addressed, delta visible at https://github.com/kubernetes/kubernetes/compare/de2ad77dc8afd3604de18cd6ae45a6644219fdb8..01a835ae5da08076c0e0cbf2c32c798884cf2d2d

**Comment by liggitt** [comment]
regenerated openapi

File: staging/src/k8s.io/api/certificates/v1beta1/types.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (smarterclayton) [reply staging/src/k8s.io/api/certificates/v1beta1/types.go:161]: Sounds reasonable to me too.

File: pkg/apis/certificates/v1beta1/defaults.go [hunk 17]
```diff
@@ -59,18 +67,34 @@ func DefaultSignerNameFromSpec(obj *certificatesv1beta1.CertificateSigningReques
 	}
 }
 
+var (
```
Comment (smarterclayton) [inline pkg/apis/certificates/v1beta1/defaults.go:70]: So much better

**Comment by smarterclayton** [comment]
API changes look fine to me, lgtm from that perspective.

**Comment by liggitt** [comment]
thanks

/hold cancelk

File: pkg/controller/certificates/signer/signer.go [hunk 18]
```diff
@@ -153,9 +154,21 @@ func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
...
+			Reason:         "SignerValidationFailure",
+			Message:        err.Error(),
+			LastUpdateTime: metav1.Now(),
+		})
+		_, err = s.client.CertificatesV1beta1().CertificateSigningRequests().UpdateStatus(context.TODO(), csr, metav1.UpdateOptions{})
+		if err != nil {
+			return fmt.Errorf("error adding failure condition for csr: %v", err)
+		}
+		return nil
+	} else if !recognized {
```
Comment (smarterclayton) [inline pkg/controller/certificates/signer/signer.go:170]: Under the assumption that another controller might be implementing those right now, but we don't need to know about it?

File: pkg/controller/certificates/signer/signer.go [hunk 19]
```diff
@@ -153,9 +154,21 @@ func (s *signer) handle(csr *capi.CertificateSigningRequest) error {
...
+		// Ignore requests for kubernetes.io signerNames we don't recognize
```
Comment (smarterclayton) [inline pkg/controller/certificates/signer/signer.go:171]: Is there a new unit test for this case?

**Comment by smarterclayton** [comment]
Only one question, I had to read the KEP again but I didn't see anything really crazy.

File: pkg/controller/certificates/signer/signer.go (hunk 18 above)
Comment (liggitt) [reply pkg/controller/certificates/signer/signer.go:170]: right

File: pkg/controller/certificates/signer/signer.go (hunk 19 above)
Comment (liggitt) [reply pkg/controller/certificates/signer/signer.go:171]: there was an existing test covering this case... I added a unit test exercising the "recognized with error" case

**Pull request merged**